python/mspasspy/db/database.py

```diff
@@ -308,53 +375,726 @@ def save_data(self, mspass_object, storage_mode='gridfs', dfile=None, dir=None,
+        (1) documents with the normalization key completely missing 
+        and (2) documents where the key is present does not match any 
+        document in normalization collection.   By default this
+        function operates silently assuming the caller will 
+        create a readable report from the return that defines 
+        the documents that had errors.  This function is used in the 
+        verify standalone program that acts as a front end to tests
+        in this module.  The function can be run in independently 
+        so there is a verbose option to print errors as they are encountered.
+        
+        :param db:  required MongoDB database handle.
+        :param wf:  mspass waveform collection on which the normalization
+        check is to be performed.  default is wf_TimeSeries.  
+        Currently only accepted alternative is wf_Seismogram.
+        :param wfquery:  optional dict passed as a query to limit the 
+        documents scanned by the function.   Default will process the 
+        entire wf collection.
+        :param verbose:  when True errors will be printed.  By default 
+        the function works silently and you should use the output to 
+        interact with any errors returned.  
+        :param error_limit: Is a sanity check on the number of errors logged.
+        Errors of any type are limited to this number (default 1000).
+        The idea is errors should be rare and if this number is exceeded 
+        you have a big problem you need to fix before scanning again.  
+        The number should be large enough to catch all condition but 
+        not so huge it become cumbersome.  With no limit or a memory 
+        fault is even possible on a huge dataset.
+        :return:  returns a tuple with two lists.  Both lists are ObjectIds
+        of the scanned wf collection that have errors.  component 0 
+        of the tuple contains ids of wf entries that have the normalization 
+        id set but the id does not resolve with the normalization collection.
+        component 1 contains the ids of documents in the wf collection that
+        do not contain the normalization id key at all (a more common problem)
 
-    def update_metadata(self, mspass_object, include_undefined=False, exclude_keys=[], collection=None):
+        """
+        # schema doesn't currently have a way to list normalized 
+        # collection names.  For now we just freeze the names 
+        # and put them in this one place for maintainability
+        norm_collection_list=['site','channel','source']
+        wf_collection_list=['wf_TimeSeries','wf_Seismogram']
+
+        if not (normalize in norm_collection_list):
+            raise MsPASSError('check_links:  illegal value for normalize arg='+normalize,
+                            'Fatal')
+        if not (wf in wf_collection_list):
+            raise MsPASSError('check_links:  illegal value for wf arg='+wf,
```

Comment:
The `wf_collection_list` and its checking is pointless here. Instead, we should only check whether `wf` is in the database or not. This should be done similar to the `read_data` method:
```
        try:
            wf_collection = self.database_schema.default_name(collection)
        except MsPASSError as err:
            raise MsPASSError('collection {} is not defined in database schema'.format(collection), 'Invalid') from err
```
Well, probably this method should be defined as:
```
def check_links(self, normalize='site', collection="wf", query={}, verbose=False, error_limit=1000):
```

The `norm_collection_list` should also be removed. Instead, we can implement a `xref_keys` method in the database schema just like the existing `required_keys` method. We can build the normalized collection list by removing the `'_id'` part of the xref_keys. 

Actually, thinking more about it, I believe this `check_links` method should be very generic that it can check links for any collections with xref_keys defined. By applying the changes above, it should already be able to do that.